File: src/main/java/ru/ezhov/rocket/action/types/OpenUrlWithTextRocketActionUi.java
package ru.ezhov.rocket.action.types;

import ru.ezhov.rocket.action.api.RocketActionConfigurationProperty;
import ru.ezhov.rocket.action.api.RocketActionSettings;
import ru.ezhov.rocket.action.icon.IconRepositoryFactory;
import ru.ezhov.rocket.action.icon.IconService;

import javax.swing.BoxLayout;
import javax.swing.JLabel;
import javax.swing.JMenu;
import javax.swing.JPanel;
import javax.swing.JTextField;
import java.awt.Color;
import java.awt.Component;
import java.awt.Desktop;
import java.awt.Graphics;
import java.net.URI;
import java.util.Arrays;
import java.util.List;
import java.util.Optional;

public class OpenUrlWithTextRocketActionUi extends AbstractRocketAction {

    private final String LABEL = "label";
    private final String DESCRIPTION = "description";
    private final String BASE_URL = "baseUrl";
    private final String PLACEHOLDER = "placeholder";
    private static final String ICON_URL = "iconUrl";

    public Component create(RocketActionSettings settings) {
        JMenu menu = new JMenu(ConfigurationUtil.getValue(settings.settings(), LABEL));
        menu.setIcon(
                IconService.load(
                        Optional.ofNullable(settings.settings().get(ICON_URL)),
                        IconRepositoryFactory.getInstance().by("link-intact-2x").get()
                )
        );

        JPanel panel = new JPanel();
        panel.setLayout(new BoxLayout(panel, BoxLayout.LINE_AXIS));
        panel.add(
                new JLabel(IconRepositoryFactory.getInstance().by("link-intact-2x").get())
        );
        TextFieldWithText textField = new TextFieldWithText(ConfigurationUtil.getValue(settings.settings(), LABEL));
        textField.setColumns(10);
        panel.add(textField);

        textField.setToolTipText(ConfigurationUtil.getValue(settings.settings(), DESCRIPTION));

        textField.addActionListener(e -> {
            String text = textField.getText();
            if (text != null && !"".equals(text)) {
                if (Desktop.isDesktopSupported()) {
                    try {
                        Desktop.getDesktop().browse(
                                new URI(
                                        ConfigurationUtil.getValue(settings.settings(), BASE_URL).replaceAll(
                                                ConfigurationUtil.getValue(settings.settings(), PLACEHOLDER), text
                                        )
                                )
                        );
                    } catch (Exception ex) {
                        ex.printStackTrace();
                    }
                }
            }
        });

        menu.add(textField);

        return menu;
    }

    @Override
    public String type() {
        return "OPEN_URL_WITH_TEXT";
    }

    @Override
    public String description() {
        return "description";
    }

    @Override
    public List<RocketActionConfigurationProperty> properties() {
        return Arrays.asList(
                createRocketActionProperty(LABEL, "TEST", true),
                createRocketActionProperty(DESCRIPTION, "TEST", true),
                createRocketActionProperty(BASE_URL, "TEST", true),
                createRocketActionProperty(PLACEHOLDER, "TEST", true),
                createRocketActionProperty(ICON_URL, "Icon URL", false)
        );
    }

    private static class TextFieldWithText extends JTextField {
        private final String text;

        public TextFieldWithText(String text) {
            this.text = text;
        }

        @Override
        public void paint(Graphics g) {
            super.paint(g);
            if ("".equals(this.getText())) {
                g.setColor(Color.gray);
                g.drawString(text, 5, this.getHeight() - 5);
            }
        }
    }
}

File: src/main/kotlin/ru/ezhov/rocket/action/types/OpenUrlWithTextRocketActionUi.kt
package ru.ezhov.rocket.action.types

import ru.ezhov.rocket.action.api.RocketActionConfigurationProperty
import ru.ezhov.rocket.action.api.RocketActionSettings
import ru.ezhov.rocket.action.icon.AppIcon
import ru.ezhov.rocket.action.icon.IconRepositoryFactory
import ru.ezhov.rocket.action.icon.IconService
import ru.ezhov.rocket.action.notification.NotificationFactory
import ru.ezhov.rocket.action.notification.NotificationType
import ru.ezhov.rocket.action.ui.swing.common.TextFieldWithText
import java.awt.Component
import java.awt.Desktop
import java.net.URI
import javax.swing.BoxLayout
import javax.swing.JLabel
import javax.swing.JMenu
import javax.swing.JPanel

class OpenUrlWithTextRocketActionUi : AbstractRocketAction() {
    override fun create(settings: RocketActionSettings): Component {
        val menu = JMenu(ConfigurationUtil.getValue(settings.settings(), LABEL))
        menu.icon = IconService().load(
                settings.settings()[ICON_URL].orEmpty(),
                IconRepositoryFactory.repository.by(AppIcon.LINK_INTACT)
        )
        val panel = JPanel()
        panel.layout = BoxLayout(panel, BoxLayout.LINE_AXIS)
        panel.add(
                JLabel(IconRepositoryFactory.repository.by(AppIcon.LINK_INTACT))
        )
        val textField = TextFieldWithText(ConfigurationUtil.getValue(settings.settings(), LABEL))
        textField.columns = 10
        panel.add(textField)
        textField.toolTipText = ConfigurationUtil.getValue(settings.settings(), DESCRIPTION)
        textField.addActionListener {
            textField
                    .text
                    ?.takeIf { it.isNotEmpty() }
                    ?.let { t ->
                        if (Desktop.isDesktopSupported()) {
                            try {
                                Desktop.getDesktop().browse(
                                        URI(
                                                ConfigurationUtil.getValue(settings.settings(), BASE_URL).replace(
                                                        ConfigurationUtil.getValue(settings.settings(), PLACEHOLDER).toRegex(), t
                                                )
                                        )
                                )
                            } catch (ex: Exception) {
                                ex.printStackTrace()
                                NotificationFactory.notification.show(NotificationType.ERROR, "Error open URL")
                            }
                        }
                    }
        }
        menu.add(textField)
        return menu
    }

    override fun type(): String {
        return "OPEN_URL_WITH_TEXT"
    }

    override fun description(): String {
        return "description"
    }

    override fun properties(): List<RocketActionConfigurationProperty> {
        return listOf(
                createRocketActionProperty(LABEL, LABEL, "TEST", true),
                createRocketActionProperty(DESCRIPTION, DESCRIPTION, "TEST", true),
                createRocketActionProperty(BASE_URL, BASE_URL, "TEST", true),
                createRocketActionProperty(PLACEHOLDER, PLACEHOLDER, "TEST", true),
                createRocketActionProperty(ICON_URL, ICON_URL, "Icon URL", false)
        )
    }

    override fun name(): String = "Открыть ссылку с подстановкой"

    companion object {
        private const val ICON_URL = "iconUrl"
        private const val LABEL = "label"
        private const val DESCRIPTION = "description"
        private const val BASE_URL = "baseUrl"
        private const val PLACEHOLDER = "placeholder"
    }
}